blockstack-vm/src/parser/mod.rs

```diff
@@ -0,0 +1,141 @@
+use errors::Error;
+use representations::SymbolicExpression;
+
+#[derive(Debug)]
+pub enum LexItem {
+    LeftParen,
+    RightParen,
+    NameParameter(String),
+    Atom(String)
+}
+
+fn finish_atom(current: &mut Option<String>) -> Option<LexItem> {
+    let resp = match current {
+        &mut None => {
+            None
+        },
+        &mut Some(ref value) => {
+            if value.starts_with('#') {
+                Some(LexItem::NameParameter(value[1..].to_string()))
+            } else {
+                Some(LexItem::Atom(value.clone()))
+            }
+        },
+    };
+
+    *current = None;
+    resp
+}
+
+pub fn lex(input: &str) -> Result<Vec<LexItem>, Error> {
+    let mut result = Vec::new();
+    let current = &mut None;
+    for c in input.chars() {
+        match c {
+            '(' => {
+                if let Some(value) = finish_atom(current) {
+                    result.push(value);
+                }
+                result.push(LexItem::LeftParen)
+            },
+            ')' => {
+                if let Some(value) = finish_atom(current) {
+                    result.push(value);
+                }
+                result.push(LexItem::RightParen)
+            },
+            '#' => {
+                if let Some(ref _value) = *current {
+                    return Err(Error::ParseError("You may not use # in the middle of an atom.".to_string()))
+                } else {
+                    *current = Some(c.to_string())
+                }
+            },
+            ' '|'\t'|'\n'|'\r' => {
+                if let Some(value) = finish_atom(current) {
+                    result.push(value);
+                }
+            },
+            _ => {
```

Comment:
Do we want to accept non-printable characters with this match arm?  Or should we just reject those as parse errors?  I feel like our smart contract VM shouldn't accept non-printable characters (and should probably _not_ use unicode) in order to ensure that the source code is non-ambiguous and doesn't have any homoglyph attacks or nefarious VT100 control codes embedded within it (which could manifest if you `cat`'ed the smart contract code to stdout in a terminal, for example).